velox/exec/TopNRowNumber.cpp

```diff
@@ -0,0 +1,280 @@
+    newRow = data_->initializeRow(topRow, true /* reuse */);
+  }
+
+  for (auto col = 0; col < input->childrenSize(); ++col) {
+    data_->store(decodedVectors_[col], index, newRow, col);
+  }
+
+  topRows.push(newRow);
+}
+
+void TopNRowNumber::noMoreInput() {
+  Operator::noMoreInput();
+
+  auto rowSize = data_->estimateRowSize();
+  if (rowSize && generateRowNumber_) {
+    rowSize.value() += sizeof(int64_t);
+  }
+
+  outputBatchSize_ = outputBatchRows(rowSize);
+  outputRows_.resize(outputBatchSize_);
+}
+
+TopNRowNumber::TopRows* TopNRowNumber::nextPartition() {
+  if (!table_) {
+    if (!currentPartition_) {
+      currentPartition_ = 0;
+      return singlePartition_.get();
+    }
+    return nullptr;
+  }
+
+  if (!currentPartition_) {
+    numPartitions_ = table_->listAllRows(
+        &partitionIt_,
+        partitions_.size(),
+        RowContainer::kUnlimited,
+        partitions_.data());
+    if (numPartitions_ == 0) {
+      // No more partitions.
+      return nullptr;
+    }
+
+    currentPartition_ = 0;
+  } else {
+    ++currentPartition_.value();
+    if (currentPartition_ >= numPartitions_) {
+      currentPartition_.reset();
+      return nextPartition();
+    }
+  }
+
+  return &currentPartition();
+}
+
+TopNRowNumber::TopRows& TopNRowNumber::currentPartition() {
+  VELOX_CHECK(currentPartition_.has_value());
+
+  if (!table_) {
+    return *singlePartition_;
+  }
+
+  return partitionAt(partitions_[currentPartition_.value()]);
+}
+
+void TopNRowNumber::appendPartitionRows(
+    TopRows& partition,
+    vector_size_t start,
+    vector_size_t size,
+    vector_size_t outputOffset,
+    FlatVector<int64_t>* rowNumbers) {
+  // Append 'size' partition rows in reverse order starting from 'start' row.
+  auto rowNumber = partition.rows.size() - start;
+  for (auto i = 0; i < size; ++i) {
+    const auto index = outputOffset + size - i - 1;
+    if (rowNumbers) {
+      rowNumbers->set(index, rowNumber--);
+    }
+    outputRows_[index] = partition.rows.top();
+    partition.rows.pop();
+  }
+}
+
+RowVectorPtr TopNRowNumber::getOutput() {
+  if (finished_ || !noMoreInput_) {
+    return nullptr;
+  }
+
+  // Loop over partitions and emit sorted rows along with row numbers.
+  auto output =
+      BaseVector::create<RowVector>(outputType_, outputBatchSize_, pool());
+  FlatVector<int64_t>* rowNumbers = nullptr;
+  if (generateRowNumber_) {
+    rowNumbers = output->children().back()->as<FlatVector<int64_t>>();
+    rowNumbers->resize(outputBatchSize_);
+  }
+
+  vector_size_t offset = 0;
+  if (remainingRowsInPartition_ > 0) {
+    auto& partition = currentPartition();
+    auto start = partition.rows.size() - remainingRowsInPartition_;
+    auto numRows =
+        std::min<vector_size_t>(outputBatchSize_, remainingRowsInPartition_);
+    appendPartitionRows(partition, start, numRows, offset, rowNumbers);
+    offset += numRows;
+    remainingRowsInPartition_ -= numRows;
+  }
+
+  while (offset < outputBatchSize_) {
```

Comment:
Is it possible to merge if branch at l226 into the while loop at l236? Thanks!

Reply:
I tried to do that, but couldn't find a nice way to do so.